Panache REST resource returns 500 - Internal Server Error for violated primary key constraints

**Describe the bug**
Having a JPA entity with a natural ID, when performing an HTTP POST that violates the primary key, a 500 error is returned along with the internal stacktrace.

**Expected behavior**
The REST resource should handle violated primary key constraint errors and return a different HTTP code (409 - Conflict may be a good choice)

**Actual behavior**
The following output is displayed: 

```bash
ᐅ http POST localhost:8080/category name=foo3 description=bar2

HTTP/1.1 500 Internal Server Error
content-length: 10911
content-type: application/json; charset=utf-8

{
    "details": "Error handling 5ebf531b-4dc5-425a-94d3-08ea0a3791a2-2, org.jboss.resteasy.spi.UnhandledException: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method",
    "stack": "org.jboss.resteasy.spi.UnhandledException: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method\n\tat org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)\n\tat org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)\n\tat org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)\n\tat org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)\n\tat org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)\n\tat org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)\n\tat org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)\n\tat org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)\n\tat org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)\n\tat io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)\n\tat io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)\n\tat io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:41)\n\tat io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)\n\tat io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\n\tat org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)\n\tat org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)\n\tat org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)\n\tat org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\n\tat org.jboss.threads.JBossThread.run(JBossThread.java:501)\nCaused by: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method\n\tat io.quarkus.registry.model.CategoryResourceJaxRs_859b58f732cd417fbbdc8447499c65f578415fa6_Subclass.add(CategoryResourceJaxRs_859b58f732cd417fbbdc8447499c65f578415fa6_Subclass.zig:484)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)\n\tat org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)\n\tat org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:643)\n\tat org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:507)\n\tat org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:457)\n\tat org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)\n\tat org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:459)\n\tat org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:419)\n\tat org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:393)\n\tat org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)\n\tat org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)\n\t... 18 more\nCaused by: javax.transaction.RollbackException: ARJUNA016053: Could not commit transaction.\n\tat com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1307)\n\tat com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:126)\n\tat io.quarkus.narayana.jta.runtime.CDIDelegatingTransactionManager.commit(CDIDelegatingTransactionManager.java:97)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.endTransaction(TransactionalInterceptorBase.java:312)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:160)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:100)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:32)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:53)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:26)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(TransactionalInterceptorRequired_Bean.zig:340)\n\tat io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)\n\tat io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)\n\tat io.quarkus.registry.model.CategoryResourceJaxRs_859b58f732cd417fbbdc8447499c65f578415fa6_Subclass.add(CategoryResourceJaxRs_859b58f732cd417fbbdc8447499c65f578415fa6_Subclass.zig:462)\n\t... 33 more\nCaused by: javax.persistence.PersistenceException: org.hibernate.exception.ConstraintViolationException: could not execute statement\n\tat org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)\n\tat org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)\n\tat org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)\n\tat org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1366)\n\tat org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:453)\n\tat org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3212)\n\tat org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2380)\n\tat org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:447)\n\tat org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:355)\n\tat org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)\n\tat org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)\n\tat com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:76)\n\tat com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:360)\n\tat com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:91)\n\tat com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:162)\n\tat com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1295)\n\t... 46 more\nCaused by: org.hibernate.exception.ConstraintViolationException: could not execute statement\n\tat org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:109)\n\tat org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)\n\tat org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)\n\tat org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)\n\tat org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200)\n\tat org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3302)\n\tat org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3829)\n\tat org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107)\n\tat org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)\n\tat org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)\n\tat java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)\n\tat org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)\n\tat org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:348)\n\tat org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)\n\tat org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:102)\n\tat org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1362)\n\t... 58 more\nCaused by: org.h2.jdbc.JdbcSQLException: Unique index or primary key violation: \"PRIMARY_KEY_3 ON PUBLIC.CATEGORY(NAME) VALUES ('foo3', 1)\"; SQL statement:\ninsert into Category (createdAt, description, name) values (?, ?, ?) [23505-197]\n\tat org.h2.message.DbException.getJdbcSQLException(DbException.java:357)\n\tat org.h2.message.DbException.get(DbException.java:179)\n\tat org.h2.message.DbException.get(DbException.java:155)\n\tat org.h2.index.BaseIndex.getDuplicateKeyException(BaseIndex.java:101)\n\tat org.h2.mvstore.db.MVSecondaryIndex.requireUnique(MVSecondaryIndex.java:236)\n\tat org.h2.mvstore.db.MVSecondaryIndex.add(MVSecondaryIndex.java:202)\n\tat org.h2.mvstore.db.MVTable.addRow(MVTable.java:732)\n\tat org.h2.command.dml.Insert.insertRows(Insert.java:182)\n\tat org.h2.command.dml.Insert.update(Insert.java:134)\n\tat org.h2.command.CommandContainer.update(CommandContainer.java:102)\n\tat org.h2.command.Command.executeUpdate(Command.java:261)\n\tat org.h2.jdbc.JdbcPreparedStatement.executeUpdateInternal(JdbcPreparedStatement.java:199)\n\tat org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:153)\n\tat io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:86)\n\tat org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)\n\t... 69 more"
}
```


**To Reproduce**

Reproducer project here: 

[category-crud.zip](https://github.com/quarkusio/quarkus/files/5808875/category-crud.zip)

Once the application is running, open a Terminal and run the following command twice: 

```
http POST localhost:8080/category name=foo description=bar
```

And you should see the error above.

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Fedora 33
 - Output of `java -version`: JDK 11
 - GraalVM version (if different from Java): Not used
 - Quarkus version or git rev: 1.11.0.CR1
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Maven

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/main/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/HibernateOrmPanacheRestProcessor.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/main/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/HibernateOrmPanacheRestProcessor.java
@@ -20,8 +20,10 @@ import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.gizmo.ClassOutput;
 import io.quarkus.hibernate.orm.rest.data.panache.PanacheEntityResource;
 import io.quarkus.hibernate.orm.rest.data.panache.PanacheRepositoryResource;
+import io.quarkus.hibernate.orm.rest.data.panache.runtime.RestDataPanacheExceptionMapper;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.RestDataResourceBuildItem;
+import io.quarkus.resteasy.common.spi.ResteasyJaxrsProviderBuildItem;
 
 class HibernateOrmPanacheRestProcessor {
 
@@ -34,6 +36,11 @@ class HibernateOrmPanacheRestProcessor {
     @BuildStep
     FeatureBuildItem feature() {
         return new FeatureBuildItem(HIBERNATE_ORM_REST_DATA_PANACHE);
+    }
+
+    @BuildStep
+    ResteasyJaxrsProviderBuildItem registerRestDataPanacheExceptionMapper() {
+        return new ResteasyJaxrsProviderBuildItem(RestDataPanacheExceptionMapper.class.getName());
     }
 
     /**

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/main/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/ResourceImplementor.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/main/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/ResourceImplementor.java
@@ -3,6 +3,7 @@ package io.quarkus.hibernate.orm.rest.data.panache.deployment;
 import java.util.List;
 
 import javax.enterprise.context.ApplicationScoped;
+import javax.transaction.Transactional;
 
 import org.jboss.jandex.FieldInfo;
 import org.jboss.logging.Logger;
@@ -89,6 +90,7 @@ class ResourceImplementor {
 
     private void implementAdd(ClassCreator classCreator, DataAccessImplementor dataAccessImplementor) {
         MethodCreator methodCreator = classCreator.getMethodCreator("add", Object.class, Object.class);
+        methodCreator.addAnnotation(Transactional.class);
         ResultHandle entity = methodCreator.getMethodParam(0);
         methodCreator.returnValue(dataAccessImplementor.persist(methodCreator, entity));
         methodCreator.close();
@@ -97,6 +99,7 @@ class ResourceImplementor {
     private void implementUpdate(ClassCreator classCreator, DataAccessImplementor dataAccessImplementor,
             String entityType) {
         MethodCreator methodCreator = classCreator.getMethodCreator("update", Object.class, Object.class, Object.class);
+        methodCreator.addAnnotation(Transactional.class);
         ResultHandle id = methodCreator.getMethodParam(0);
         ResultHandle entity = methodCreator.getMethodParam(1);
         // Set entity ID before executing an update to make sure that a requested object ID matches a given entity ID.
@@ -107,6 +110,7 @@ class ResourceImplementor {
 
     private void implementDelete(ClassCreator classCreator, DataAccessImplementor dataAccessImplementor) {
         MethodCreator methodCreator = classCreator.getMethodCreator("delete", boolean.class, Object.class);
+        methodCreator.addAnnotation(Transactional.class);
         ResultHandle id = methodCreator.getMethodParam(0);
         methodCreator.returnValue(dataAccessImplementor.deleteById(methodCreator, id));
         methodCreator.close();

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/AbstractPostMethodTest.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/AbstractPostMethodTest.java
@@ -60,6 +60,11 @@ public abstract class AbstractPostMethodTest {
                 .and().body("id", is(equalTo("test-complex")))
                 .and().body("name", is(equalTo("test collection")))
                 .and().body("items", is(empty()));
+        given().accept("application/json")
+                .and().contentType("application/json")
+                .and().body("{\"id\": \"test-complex\", \"name\": \"test collection\"}")
+                .when().post("/collections")
+                .then().statusCode(409);
     }
 
     @Test
@@ -78,5 +83,10 @@ public abstract class AbstractPostMethodTest {
                 .and().body("_links.self.href", endsWith("/collections/test-complex-hal"))
                 .and().body("_links.update.href", endsWith("/collections/test-complex-hal"))
                 .and().body("_links.remove.href", endsWith("/collections/test-complex-hal"));
+        given().accept("application/hal+json")
+                .and().contentType("application/json")
+                .and().body("{\"id\": \"test-complex-hal\", \"name\": \"test collection\"}")
+                .when().post("/collections")
+                .then().statusCode(409);
     }
 }

--- a/extensions/panache/hibernate-orm-rest-data-panache/runtime/src/main/java/io/quarkus/hibernate/orm/rest/data/panache/runtime/RestDataPanacheExceptionMapper.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/runtime/src/main/java/io/quarkus/hibernate/orm/rest/data/panache/runtime/RestDataPanacheExceptionMapper.java
@@ -1,0 +1,48 @@
+package io.quarkus.hibernate.orm.rest.data.panache.runtime;
+
+import javax.persistence.PersistenceException;
+import javax.transaction.RollbackException;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+
+import org.hibernate.exception.ConstraintViolationException;
+
+import io.quarkus.arc.ArcUndeclaredThrowableException;
+import io.quarkus.rest.data.panache.RestDataPanacheException;
+
+public class RestDataPanacheExceptionMapper implements ExceptionMapper<RestDataPanacheException> {
+    @Override
+    public Response toResponse(RestDataPanacheException exception) {
+        exception.printStackTrace();
+
+        if (exception.getCause() instanceof ArcUndeclaredThrowableException) {
+            return toResponse((ArcUndeclaredThrowableException) exception.getCause(), exception.getMessage());
+        }
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), exception.getMessage()).build();
+    }
+
+    private Response toResponse(ArcUndeclaredThrowableException exception, String message) {
+        if (exception.getCause() instanceof RollbackException) {
+            return toResponse((RollbackException) exception.getCause(), message);
+        }
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), message).build();
+    }
+
+    private Response toResponse(RollbackException exception, String message) {
+        if (exception.getCause() instanceof PersistenceException) {
+            return toResponse((PersistenceException) exception.getCause(), message);
+        }
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), message).build();
+    }
+
+    private Response toResponse(PersistenceException exception, String message) {
+        if (exception.getCause() instanceof ConstraintViolationException) {
+            return toResponse((ConstraintViolationException) exception.getCause(), message);
+        }
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), message).build();
+    }
+
+    private Response toResponse(ConstraintViolationException exception, String message) {
+        return Response.status(Response.Status.CONFLICT.getStatusCode(), message).build();
+    }
+}

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/AddMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/AddMethodImplementor.java
@@ -8,6 +8,7 @@ import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.rest.data.panache.RestDataResource;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.properties.ResourceProperties;
@@ -27,7 +28,6 @@ public final class AddMethodImplementor extends StandardMethodImplementor {
      *
      * <pre>
      * {@code
-     *     &#64;Transactional
      *     &#64;POST
      *     &#64;Path("")
      *     &#64;Consumes({"application/json"})
@@ -37,15 +37,19 @@ public final class AddMethodImplementor extends StandardMethodImplementor {
      *         entityClassName = "com.example.Entity"
      *     )
      *     public Response add(Entity entityToSave) {
-     *         Entity entity = restDataResource.add(entityToSave);
-     *         String location = new ResourceLinksProvider().getSelfLink(entity);
-     *         if (location != null) {
-     *             ResponseBuilder responseBuilder = Response.status(201);
-     *             responseBuilder.entity(entity);
-     *             responseBuilder.location(URI.create(location));
-     *             return responseBuilder.build();
-     *         } else {
-     *             throw new RuntimeException("Could not extract a new entity URL")
+     *         try {
+     *             Entity entity = restDataResource.add(entityToSave);
+     *             String location = new ResourceLinksProvider().getSelfLink(entity);
+     *             if (location != null) {
+     *                 ResponseBuilder responseBuilder = Response.status(201);
+     *                 responseBuilder.entity(entity);
+     *                 responseBuilder.location(URI.create(location));
+     *                 return responseBuilder.build();
+     *             } else {
+     *                 throw new RuntimeException("Could not extract a new entity URL")
+     *             }
+     *         } catch (Throwable t) {
+     *             throw new RestDataPanacheException(t);
      *         }
      *     }
      * }
@@ -54,26 +58,28 @@ public final class AddMethodImplementor extends StandardMethodImplementor {
     @Override
     protected void implementInternal(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceField) {
-        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class.getName(),
+        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class,
                 resourceMetadata.getEntityType());
 
         // Add method annotations
         addPathAnnotation(methodCreator, resourceProperties.getPath(RESOURCE_METHOD_NAME));
-        addTransactionalAnnotation(methodCreator);
         addPostAnnotation(methodCreator);
         addConsumesAnnotation(methodCreator, APPLICATION_JSON);
         addProducesAnnotation(methodCreator, APPLICATION_JSON);
         addLinksAnnotation(methodCreator, resourceMetadata.getEntityType(), REL);
 
-        // Invoke resource methods
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
         ResultHandle entityToSave = methodCreator.getMethodParam(0);
-        ResultHandle entity = methodCreator.invokeVirtualMethod(
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to add an entity");
+        ResultHandle entity = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, Object.class, Object.class),
                 resource, entityToSave);
-
         // Return response
-        methodCreator.returnValue(ResponseImplementor.created(methodCreator, entity));
+        tryBlock.returnValue(ResponseImplementor.created(tryBlock, entity));
+
+        tryBlock.close();
         methodCreator.close();
     }
 

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/DeleteMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/DeleteMethodImplementor.java
@@ -2,11 +2,14 @@ package io.quarkus.rest.data.panache.deployment.methods;
 
 import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
 
+import javax.ws.rs.core.Response;
+
 import io.quarkus.gizmo.BranchResult;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.rest.data.panache.RestDataResource;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.properties.ResourceProperties;
@@ -26,16 +29,22 @@ public final class DeleteMethodImplementor extends StandardMethodImplementor {
      *
      * <pre>
      * {@code
-     *     &#64;Transactional
      *     &#64;DELETE
      *     &#64;Path("{id}")
      *     &#64;LinkResource(
      *         rel = "remove",
      *         entityClassName = "com.example.Entity"
      *     )
-     *     public void delete(@PathParam("id") ID id) {
-     *         if (!restDataResource.delete(id)) {
-     *             throw new WebApplicationException(404);
+     *     public Response delete(@PathParam("id") ID id) {
+     *         try {
+     *             boolean deleted = restDataResource.delete(id);
+     *             if (deleted) {
+     *                 return Response.noContent().build();
+     *             } else {
+     *                 return Response.status(404).build();
+     *             }
+     *         } catch (Throwable t) {
+     *             throw new RestDataPanacheException(t);
      *         }
      *     }
      * }
@@ -44,28 +53,30 @@ public final class DeleteMethodImplementor extends StandardMethodImplementor {
     @Override
     protected void implementInternal(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceField) {
-        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, void.class.getName(),
+        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class,
                 resourceMetadata.getIdType());
 
         // Add method annotations
         addPathAnnotation(methodCreator, appendToPath(resourceProperties.getPath(RESOURCE_METHOD_NAME), "{id}"));
-        addTransactionalAnnotation(methodCreator);
         addDeleteAnnotation(methodCreator);
         addPathParamAnnotation(methodCreator.getParameterAnnotations(0), "id");
         addLinksAnnotation(methodCreator, resourceMetadata.getEntityType(), REL);
 
-        // Invoke resource methods
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
         ResultHandle id = methodCreator.getMethodParam(0);
-        ResultHandle result = methodCreator.invokeVirtualMethod(
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to delete an entity");
+        ResultHandle deleted = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, boolean.class, Object.class),
                 resource, id);
-        BranchResult entityWasDeleted = methodCreator.ifNonZero(result);
 
         // Return response
-        entityWasDeleted.trueBranch().returnValue(null);
-        entityWasDeleted.falseBranch()
-                .throwException(ResponseImplementor.notFoundException(entityWasDeleted.falseBranch()));
+        BranchResult entityWasDeleted = tryBlock.ifNonZero(deleted);
+        entityWasDeleted.trueBranch().returnValue(ResponseImplementor.noContent(entityWasDeleted.trueBranch()));
+        entityWasDeleted.falseBranch().returnValue(ResponseImplementor.notFound(entityWasDeleted.falseBranch()));
+
+        tryBlock.close();
         methodCreator.close();
     }
 

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/GetMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/GetMethodImplementor.java
@@ -2,11 +2,14 @@ package io.quarkus.rest.data.panache.deployment.methods;
 
 import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
 
+import javax.ws.rs.core.Response;
+
 import io.quarkus.gizmo.BranchResult;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.rest.data.panache.RestDataResource;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.properties.ResourceProperties;
@@ -33,12 +36,16 @@ public final class GetMethodImplementor extends StandardMethodImplementor {
      *         rel = "self",
      *         entityClassName = "com.example.Entity"
      *     )
-     *     public Entity get(@PathParam("id") ID id) {
-     *         Entity entity = restDataResource.get(id);
-     *         if (entity != null) {
-     *             return entity;
-     *         } else {
-     *             throw new WebApplicationException(404);
+     *     public Response get(@PathParam("id") ID id) {
+     *         try {
+     *             Entity entity = restDataResource.get(id);
+     *             if (entity != null) {
+     *                 return entity;
+     *             } else {
+     *                 return Response.status(404).build();
+     *             }
+     *         } catch (Throwable t) {
+     *             throw new RestDataPanacheException(t);
      *         }
      *     }
      * }
@@ -47,7 +54,7 @@ public final class GetMethodImplementor extends StandardMethodImplementor {
     @Override
     protected void implementInternal(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceField) {
-        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, resourceMetadata.getEntityType(),
+        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class,
                 resourceMetadata.getIdType());
 
         // Add method annotations
@@ -57,17 +64,21 @@ public final class GetMethodImplementor extends StandardMethodImplementor {
         addPathParamAnnotation(methodCreator.getParameterAnnotations(0), "id");
         addLinksAnnotation(methodCreator, resourceMetadata.getEntityType(), REL);
 
-        // Invoke resource methods
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
         ResultHandle id = methodCreator.getMethodParam(0);
-        ResultHandle entity = methodCreator.invokeVirtualMethod(
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to get an entity");
+        ResultHandle entity = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, Object.class, Object.class),
                 resource, id);
-        BranchResult entityNotFound = methodCreator.ifNull(entity);
 
         // Return response
-        entityNotFound.trueBranch().throwException(ResponseImplementor.notFoundException(entityNotFound.trueBranch()));
-        entityNotFound.falseBranch().returnValue(entity);
+        BranchResult wasNotFound = tryBlock.ifNull(entity);
+        wasNotFound.trueBranch().returnValue(ResponseImplementor.notFound(wasNotFound.trueBranch()));
+        wasNotFound.falseBranch().returnValue(ResponseImplementor.ok(wasNotFound.falseBranch(), entity));
+
+        tryBlock.close();
         methodCreator.close();
     }
 

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/ListMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/ListMethodImplementor.java
@@ -13,6 +13,7 @@ import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.panache.common.Page;
 import io.quarkus.panache.common.Sort;
 import io.quarkus.rest.data.panache.RestDataResource;
@@ -54,12 +55,16 @@ public final class ListMethodImplementor extends StandardMethodImplementor {
      *             &#64;QueryParam("sort") String sortQuery) {
      *         Page page = Page.of(pageIndex, pageSize);
      *         Sort sort = ...; // Build a sort instance by parsing a query param
-     *         List<Entity> entities = resource.getAll(page, sort);
-     *         // Get the page count, and build first, last, next, previous page instances
-     *         Response.ResponseBuilder responseBuilder = Response.status(200);
-     *         responseBuilder.entity(entities);
-     *         // Add headers with first, last, next and previous page URIs if they exist
-     *         return responseBuilder.build();
+     *         try {
+     *             List<Entity> entities = resource.getAll(page, sort);
+     *             // Get the page count, and build first, last, next, previous page instances
+     *             Response.ResponseBuilder responseBuilder = Response.status(200);
+     *             responseBuilder.entity(entities);
+     *             // Add headers with first, last, next and previous page URIs if they exist
+     *             return responseBuilder.build();
+     *         } catch (Throwable t) {
+     *             throw new RestDataPanacheException(t);
+     *         }
      *     }
      * }
      * </pre>
@@ -99,25 +104,28 @@ public final class ListMethodImplementor extends StandardMethodImplementor {
         addContextAnnotation(methodCreator.getParameterAnnotations(3));
 
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
-
-        // Invoke resource methods
         ResultHandle sortQuery = methodCreator.getMethodParam(0);
         ResultHandle sort = sortImplementor.getSort(methodCreator, sortQuery);
         ResultHandle pageIndex = methodCreator.getMethodParam(1);
         ResultHandle pageSize = methodCreator.getMethodParam(2);
         ResultHandle page = paginationImplementor.getPage(methodCreator, pageIndex, pageSize);
-        ResultHandle pageCount = methodCreator.invokeVirtualMethod(
+        ResultHandle uriInfo = methodCreator.getMethodParam(3);
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to list the entities");
+        ResultHandle pageCount = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), Constants.PAGE_COUNT_METHOD_PREFIX + RESOURCE_METHOD_NAME,
                         int.class, Page.class),
                 resource, page);
-        ResultHandle uriInfo = methodCreator.getMethodParam(3);
-        ResultHandle links = paginationImplementor.getLinks(methodCreator, uriInfo, page, pageCount);
-        ResultHandle entities = methodCreator.invokeVirtualMethod(
+        ResultHandle links = paginationImplementor.getLinks(tryBlock, uriInfo, page, pageCount);
+        ResultHandle entities = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, List.class, Page.class, Sort.class),
                 resource, page, sort);
 
         // Return response
-        methodCreator.returnValue(ResponseImplementor.ok(methodCreator, entities, links));
+        tryBlock.returnValue(ResponseImplementor.ok(tryBlock, entities, links));
+
+        tryBlock.close();
         methodCreator.close();
     }
 
@@ -132,16 +140,20 @@ public final class ListMethodImplementor extends StandardMethodImplementor {
         addLinksAnnotation(methodCreator, resourceMetadata.getEntityType(), REL);
         addQueryParamAnnotation(methodCreator.getParameterAnnotations(0), "sort");
 
-        // Invoke resource methods
         ResultHandle sortQuery = methodCreator.getMethodParam(0);
         ResultHandle sort = sortImplementor.getSort(methodCreator, sortQuery);
         ResultHandle resource = methodCreator.readInstanceField(resourceFieldDescriptor, methodCreator.getThis());
-        ResultHandle entities = methodCreator.invokeVirtualMethod(
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to list the entities");
+        ResultHandle entities = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, List.class, Page.class, Sort.class),
-                resource, methodCreator.loadNull(), sort);
+                resource, tryBlock.loadNull(), sort);
 
         // Return response
-        methodCreator.returnValue(ResponseImplementor.ok(methodCreator, entities));
+        tryBlock.returnValue(ResponseImplementor.ok(tryBlock, entities));
+
+        tryBlock.close();
         methodCreator.close();
     }
 }

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/StandardMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/StandardMethodImplementor.java
@@ -1,6 +1,5 @@
 package io.quarkus.rest.data.panache.deployment.methods;
 
-import javax.transaction.Transactional;
 import javax.ws.rs.Consumes;
 import javax.ws.rs.DELETE;
 import javax.ws.rs.DefaultValue;
@@ -17,8 +16,12 @@ import org.jboss.resteasy.links.LinkResource;
 
 import io.quarkus.gizmo.AnnotatedElement;
 import io.quarkus.gizmo.AnnotationCreator;
+import io.quarkus.gizmo.BytecodeCreator;
+import io.quarkus.gizmo.CatchBlockCreator;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
+import io.quarkus.gizmo.TryBlock;
+import io.quarkus.rest.data.panache.RestDataPanacheException;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.properties.ResourceProperties;
 import io.quarkus.rest.data.panache.runtime.sort.SortQueryParamValidator;
@@ -50,8 +53,12 @@ public abstract class StandardMethodImplementor implements MethodImplementor {
      */
     protected abstract String getResourceMethodName();
 
-    protected void addTransactionalAnnotation(AnnotatedElement element) {
-        element.addAnnotation(Transactional.class);
+    protected TryBlock implementTryBlock(BytecodeCreator bytecodeCreator, String message) {
+        TryBlock tryBlock = bytecodeCreator.tryBlock();
+        CatchBlockCreator catchBlock = tryBlock.addCatch(Throwable.class);
+        catchBlock.throwException(RestDataPanacheException.class, message, catchBlock.getCaughtException());
+        catchBlock.close();
+        return tryBlock;
     }
 
     protected void addGetAnnotation(AnnotatedElement element) {

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/UpdateMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/UpdateMethodImplementor.java
@@ -10,6 +10,7 @@ import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.rest.data.panache.RestDataResource;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.properties.ResourceProperties;
@@ -32,7 +33,6 @@ public final class UpdateMethodImplementor extends StandardMethodImplementor {
      *
      * <pre>
      * {@code
-     *     &#64;Transactional
      *     &#64;PUT
      *     &#64;Path("{id}")
      *     &#64;Consumes({"application/json"})
@@ -42,19 +42,23 @@ public final class UpdateMethodImplementor extends StandardMethodImplementor {
      *         entityClassName = "com.example.Entity"
      *     )
      *     public Response update(@PathParam("id") ID id, Entity entityToSave) {
-     *         if (resource.get(id) != null) {
-     *             resource.update(id, entityToSave);
-     *             return Response.status(204).build();
-     *         } else {
-     *             Entity entity = resource.update(id, entityToSave);
-     *             String location = new ResourceLinksProvider().getSelfLink(entity);
-     *             if (location != null) {
-     *                 ResponseBuilder responseBuilder = Response.status(201);
-     *                 responseBuilder.entity(entity);
-     *                 responseBuilder.location(URI.create(location));
-     *                 return responseBuilder.build();
+     *         try {
+     *             if (resource.get(id) != null) {
+     *                 resource.update(id, entityToSave);
+     *                 return Response.status(204).build();
      *             } else {
-     *                 throw new RuntimeException("Could not extract a new entity URL")
+     *                 Entity entity = resource.update(id, entityToSave);
+     *                 String location = new ResourceLinksProvider().getSelfLink(entity);
+     *                 if (location != null) {
+     *                     ResponseBuilder responseBuilder = Response.status(201);
+     *                     responseBuilder.entity(entity);
+     *                     responseBuilder.location(URI.create(location));
+     *                     return responseBuilder.build();
+     *                 } else {
+     *                     throw new RuntimeException("Could not extract a new entity URL")
+     *                 }
+     *             } catch (Throwable t) {
+     *                 throw new RestDataPanacheException(t);
      *             }
      *         }
      *     }
@@ -64,27 +68,29 @@ public final class UpdateMethodImplementor extends StandardMethodImplementor {
     @Override
     protected void implementInternal(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceField) {
-        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class.getName(),
+        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class,
                 resourceMetadata.getIdType(), resourceMetadata.getEntityType());
 
         // Add method annotations
         addPathAnnotation(methodCreator,
                 appendToPath(resourceProperties.getPath(RESOURCE_UPDATE_METHOD_NAME), "{id}"));
-        addTransactionalAnnotation(methodCreator);
         addPutAnnotation(methodCreator);
         addPathParamAnnotation(methodCreator.getParameterAnnotations(0), "id");
         addConsumesAnnotation(methodCreator, APPLICATION_JSON);
         addProducesAnnotation(methodCreator, APPLICATION_JSON);
         addLinksAnnotation(methodCreator, resourceMetadata.getEntityType(), REL);
 
-        // Invoke resource methods
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
         ResultHandle id = methodCreator.getMethodParam(0);
         ResultHandle entityToSave = methodCreator.getMethodParam(1);
 
-        BranchResult entityExists = doesEntityExist(methodCreator, resourceMetadata.getResourceClass(), resource, id);
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to update an entity");
+        BranchResult entityExists = doesEntityExist(tryBlock, resourceMetadata.getResourceClass(), resource, id);
         updateAndReturn(entityExists.trueBranch(), resourceMetadata.getResourceClass(), resource, id, entityToSave);
         createAndReturn(entityExists.falseBranch(), resourceMetadata.getResourceClass(), resource, id, entityToSave);
+
+        tryBlock.close();
         methodCreator.close();
     }
 

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/hal/AddHalMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/hal/AddHalMethodImplementor.java
@@ -8,6 +8,7 @@ import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.rest.data.panache.RestDataResource;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.properties.ResourceProperties;
@@ -25,22 +26,25 @@ public final class AddHalMethodImplementor extends HalMethodImplementor {
      *
      * <pre>
      * {@code
-     *     &#64;Transactional
      *     &#64;POST
      *     &#64;Path("")
      *     &#64;Consumes({"application/json"})
      *     &#64;Produces({"application/hal+json"})
      *     public Response addHal(Entity entityToSave) {
-     *         Entity entity = resource.add(entityToSave);
-     *         HalEntityWrapper wrapper = new HalEntityWrapper(entity);
-     *         String location = new ResourceLinksProvider().getSelfLink(entity);
-     *         if (location != null) {
-     *             ResponseBuilder responseBuilder = Response.status(201);
-     *             responseBuilder.entity(wrapper);
-     *             responseBuilder.location(URI.create(location));
-     *             return responseBuilder.build();
-     *         } else {
-     *             throw new RuntimeException("Could not extract a new entity URL")
+     *         try {
+     *             Entity entity = resource.add(entityToSave);
+     *             HalEntityWrapper wrapper = new HalEntityWrapper(entity);
+     *             String location = new ResourceLinksProvider().getSelfLink(entity);
+     *             if (location != null) {
+     *                 ResponseBuilder responseBuilder = Response.status(201);
+     *                 responseBuilder.entity(wrapper);
+     *                 responseBuilder.location(URI.create(location));
+     *                 return responseBuilder.build();
+     *             } else {
+     *                 throw new RuntimeException("Could not extract a new entity URL");
+     *             }
+     *         } catch (Throwable t) {
+     *             throw new RestDataPanacheException(t);
      *         }
      *     }
      * }
@@ -49,26 +53,29 @@ public final class AddHalMethodImplementor extends HalMethodImplementor {
     @Override
     protected void implementInternal(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceField) {
-        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class.getName(),
+        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class,
                 resourceMetadata.getEntityType());
 
         // Add method annotations
         addPathAnnotation(methodCreator, resourceProperties.getPath(RESOURCE_METHOD_NAME));
-        addTransactionalAnnotation(methodCreator);
         addPostAnnotation(methodCreator);
         addConsumesAnnotation(methodCreator, APPLICATION_JSON);
         addProducesAnnotation(methodCreator, APPLICATION_HAL_JSON);
 
-        // Invoke resource methods
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
         ResultHandle entityToSave = methodCreator.getMethodParam(0);
-        ResultHandle entity = methodCreator.invokeVirtualMethod(
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to add an entity");
+        ResultHandle entity = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, Object.class, Object.class),
                 resource, entityToSave);
 
         // Wrap and return response
-        methodCreator.returnValue(ResponseImplementor.created(methodCreator, wrapHalEntity(methodCreator, entity),
-                ResponseImplementor.getEntityUrl(methodCreator, entity)));
+        tryBlock.returnValue(ResponseImplementor.created(tryBlock, wrapHalEntity(tryBlock, entity),
+                ResponseImplementor.getEntityUrl(tryBlock, entity)));
+
+        tryBlock.close();
         methodCreator.close();
     }
 

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/hal/GetHalMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/hal/GetHalMethodImplementor.java
@@ -2,11 +2,14 @@ package io.quarkus.rest.data.panache.deployment.methods.hal;
 
 import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
 
+import javax.ws.rs.core.Response;
+
 import io.quarkus.gizmo.BranchResult;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.rest.data.panache.RestDataResource;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.properties.ResourceProperties;
@@ -28,12 +31,16 @@ public final class GetHalMethodImplementor extends HalMethodImplementor {
      *     &#64;GET
      *     &#64;Produces({"application/hal+json"})
      *     &#64;Path("{id}")
-     *     public HalEntityWrapper getHal(@PathParam("id") ID id) {
-     *         Entity entity = resource.get(id);
-     *         if (entity != null) {
-     *             return new HalEntityWrapper(entity);
-     *         } else {
-     *             throw new WebApplicationException(404);
+     *     public Response getHal(@PathParam("id") ID id) {
+     *         try {
+     *             Entity entity = resource.get(id);
+     *             if (entity != null) {
+     *                 return Response.ok(new HalEntityWrapper(entity)).build();
+     *             } else {
+     *                 return Response.status(404).build();
+     *             }
+     *         } catch (Throwable t) {
+     *             throw new RestDataPanacheException(t);
      *         }
      *     }
      * }
@@ -42,7 +49,7 @@ public final class GetHalMethodImplementor extends HalMethodImplementor {
     @Override
     protected void implementInternal(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceField) {
-        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, HalEntityWrapper.class,
+        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class,
                 resourceMetadata.getIdType());
 
         // Add method annotations
@@ -51,17 +58,22 @@ public final class GetHalMethodImplementor extends HalMethodImplementor {
         addProducesAnnotation(methodCreator, APPLICATION_HAL_JSON);
         addPathParamAnnotation(methodCreator.getParameterAnnotations(0), "id");
 
-        // Invoke resource methods
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
         ResultHandle id = methodCreator.getMethodParam(0);
-        ResultHandle entity = methodCreator.invokeVirtualMethod(
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to get an entity");
+        ResultHandle entity = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, Object.class, Object.class),
                 resource, id);
-        BranchResult entityNotFound = methodCreator.ifNull(entity);
 
         // Wrap and return response
-        entityNotFound.trueBranch().throwException(ResponseImplementor.notFoundException(entityNotFound.trueBranch()));
-        entityNotFound.falseBranch().returnValue(wrapHalEntity(entityNotFound.falseBranch(), entity));
+        BranchResult wasNotFound = tryBlock.ifNull(entity);
+        wasNotFound.trueBranch().returnValue(ResponseImplementor.notFound(wasNotFound.trueBranch()));
+        wasNotFound.falseBranch().returnValue(
+                ResponseImplementor.ok(wasNotFound.falseBranch(), wrapHalEntity(wasNotFound.falseBranch(), entity)));
+
+        tryBlock.close();
         methodCreator.close();
     }
 

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/hal/ListHalMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/hal/ListHalMethodImplementor.java
@@ -14,6 +14,7 @@ import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.panache.common.Page;
 import io.quarkus.panache.common.Sort;
 import io.quarkus.rest.data.panache.RestDataResource;
@@ -50,14 +51,18 @@ public final class ListHalMethodImplementor extends HalMethodImplementor {
      *             &#64;QueryParam("sort") String sortQuery) {
      *         Page page = Page.of(pageIndex, pageSize);
      *         Sort sort = ...; // Build a sort instance by parsing a query param
-     *         List<Entity> entities = resource.getAll(page, sort);
-     *         // Get the page count, and build first, last, next, previous page instances
-     *         HalCollectionWrapper wrapper = new HalCollectionWrapper(entities, Entity.class, "entities");
-     *         // Add first, last, next and previous page URIs to the wrapper if they exist
-     *         Response.ResponseBuilder responseBuilder = Response.status(200);
-     *         responseBuilder.entity(wrapper);
-     *         // Add headers with first, last, next and previous page URIs if they exist
-     *         return responseBuilder.build();
+     *         try {
+     *             List<Entity> entities = resource.getAll(page, sort);
+     *             // Get the page count, and build first, last, next, previous page instances
+     *             HalCollectionWrapper wrapper = new HalCollectionWrapper(entities, Entity.class, "entities");
+     *             // Add first, last, next and previous page URIs to the wrapper if they exist
+     *             Response.ResponseBuilder responseBuilder = Response.status(200);
+     *             responseBuilder.entity(wrapper);
+     *             // Add headers with first, last, next and previous page URIs if they exist
+     *             return responseBuilder.build();
+     *         } catch (Throwable t) {
+     *             throw new RestDataPanacheException(t);
+     *         }
      *    }
      * }
      * </pre>
@@ -80,7 +85,6 @@ public final class ListHalMethodImplementor extends HalMethodImplementor {
     private void implementPaged(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceField) {
         // Method parameters: sort strings, page index, page size, uri info
-        // TODO could list be an Iterator?
         MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class, List.class, int.class,
                 int.class, UriInfo.class);
 
@@ -96,36 +100,38 @@ public final class ListHalMethodImplementor extends HalMethodImplementor {
         addDefaultValueAnnotation(methodCreator.getParameterAnnotations(2), Integer.toString(DEFAULT_PAGE_SIZE));
         addContextAnnotation(methodCreator.getParameterAnnotations(3));
 
-        // Invoke resource methods
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
-
         ResultHandle sortQuery = methodCreator.getMethodParam(0);
         ResultHandle sort = sortImplementor.getSort(methodCreator, sortQuery);
         ResultHandle pageIndex = methodCreator.getMethodParam(1);
         ResultHandle pageSize = methodCreator.getMethodParam(2);
         ResultHandle page = paginationImplementor.getPage(methodCreator, pageIndex, pageSize);
-        ResultHandle pageCount = methodCreator.invokeVirtualMethod(
+        ResultHandle uriInfo = methodCreator.getMethodParam(3);
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to list the entities");
+        ResultHandle pageCount = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), Constants.PAGE_COUNT_METHOD_PREFIX + RESOURCE_METHOD_NAME,
                         int.class, Page.class),
                 resource, page);
-        ResultHandle uriInfo = methodCreator.getMethodParam(3);
-        ResultHandle links = paginationImplementor.getLinks(methodCreator, uriInfo, page, pageCount);
-        ResultHandle entities = methodCreator.invokeVirtualMethod(
+        ResultHandle links = paginationImplementor.getLinks(tryBlock, uriInfo, page, pageCount);
+        ResultHandle entities = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, List.class, Page.class, Sort.class),
                 resource, page, sort);
 
         // Wrap and return response
-        ResultHandle wrapper = wrapHalEntities(methodCreator, entities, resourceMetadata.getEntityType(),
+        ResultHandle wrapper = wrapHalEntities(tryBlock, entities, resourceMetadata.getEntityType(),
                 resourceProperties.getHalCollectionName());
-        methodCreator.invokeVirtualMethod(
+        tryBlock.invokeVirtualMethod(
                 ofMethod(HalCollectionWrapper.class, "addLinks", void.class, Link[].class), wrapper, links);
-        methodCreator.returnValue(ResponseImplementor.ok(methodCreator, wrapper, links));
+        tryBlock.returnValue(ResponseImplementor.ok(tryBlock, wrapper, links));
+
+        tryBlock.close();
         methodCreator.close();
     }
 
     private void implementNotPaged(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceFieldDescriptor) {
-        // TODO could list be an Iterator?
         MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class, List.class);
 
         // Add method annotations
@@ -134,18 +140,22 @@ public final class ListHalMethodImplementor extends HalMethodImplementor {
         addProducesAnnotation(methodCreator, APPLICATION_HAL_JSON);
         addQueryParamAnnotation(methodCreator.getParameterAnnotations(0), "sort");
 
-        // Invoke resource methods
         ResultHandle sortQuery = methodCreator.getMethodParam(0);
         ResultHandle sort = sortImplementor.getSort(methodCreator, sortQuery);
         ResultHandle resource = methodCreator.readInstanceField(resourceFieldDescriptor, methodCreator.getThis());
-        ResultHandle entities = methodCreator.invokeVirtualMethod(
+
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to list the entities");
+        ResultHandle entities = tryBlock.invokeVirtualMethod(
                 ofMethod(resourceMetadata.getResourceClass(), RESOURCE_METHOD_NAME, List.class, Page.class, Sort.class),
-                resource, methodCreator.loadNull(), sort);
+                resource, tryBlock.loadNull(), sort);
 
         // Wrap and return response
-        ResultHandle wrapper = wrapHalEntities(methodCreator, entities, resourceMetadata.getEntityType(),
+        ResultHandle wrapper = wrapHalEntities(tryBlock, entities, resourceMetadata.getEntityType(),
                 resourceProperties.getHalCollectionName());
-        methodCreator.returnValue(ResponseImplementor.ok(methodCreator, wrapper));
+        tryBlock.returnValue(ResponseImplementor.ok(tryBlock, wrapper));
+
+        tryBlock.close();
         methodCreator.close();
     }
 }

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/hal/UpdateHalMethodImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/hal/UpdateHalMethodImplementor.java
@@ -10,6 +10,7 @@ import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.gizmo.TryBlock;
 import io.quarkus.rest.data.panache.RestDataResource;
 import io.quarkus.rest.data.panache.deployment.ResourceMetadata;
 import io.quarkus.rest.data.panache.deployment.properties.ResourceProperties;
@@ -29,27 +30,30 @@ public final class UpdateHalMethodImplementor extends HalMethodImplementor {
      *
      * <pre>
      * {@code
-     *     &#64;Transactional
      *     &#64;PUT
      *     &#64;Path("{id}")
      *     &#64;Consumes({"application/json"})
      *     &#64;Produces({"application/hal+json"})
      *     public Response updateHal(@PathParam("id") ID id, Entity entityToSave) {
-     *         if (resource.get(id) != null) {
-     *             resource.update(id, entityToSave);
-     *             return Response.status(204).build();
-     *         } else {
-     *             Entity entity = resource.update(id, entityToSave);
-     *             HalEntityWrapper wrapper = new HalEntityWrapper(entity);
-     *             String location = new ResourceLinksProvider().getSelfLink(entity);
-     *             if (location != null) {
-     *                 ResponseBuilder responseBuilder = Response.status(201);
-     *                 responseBuilder.entity(wrapper);
-     *                 responseBuilder.location(URI.create(location));
-     *                 return responseBuilder.build();
+     *         try {
+     *             if (resource.get(id) != null) {
+     *                 resource.update(id, entityToSave);
+     *                 return Response.status(204).build();
      *             } else {
-     *                 throw new RuntimeException("Could not extract a new entity URL")
-     *             }
+     *                 Entity entity = resource.update(id, entityToSave);
+     *                 HalEntityWrapper wrapper = new HalEntityWrapper(entity);
+     *                 String location = new ResourceLinksProvider().getSelfLink(entity);
+     *                 if (location != null) {
+     *                     ResponseBuilder responseBuilder = Response.status(201);
+     *                     responseBuilder.entity(wrapper);
+     *                     responseBuilder.location(URI.create(location));
+     *                     return responseBuilder.build();
+     *                 } else {
+     *                     throw new RuntimeException("Could not extract a new entity URL")
+     *                 }
+     *              }
+     *         } catch (Throwable t) {
+     *             throw new RestDataPanacheException(t);
      *         }
      *     }
      * }
@@ -58,27 +62,28 @@ public final class UpdateHalMethodImplementor extends HalMethodImplementor {
     @Override
     protected void implementInternal(ClassCreator classCreator, ResourceMetadata resourceMetadata,
             ResourceProperties resourceProperties, FieldDescriptor resourceField) {
-        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class.getName(),
+        MethodCreator methodCreator = classCreator.getMethodCreator(METHOD_NAME, Response.class,
                 resourceMetadata.getIdType(), resourceMetadata.getEntityType());
 
         // Add method annotations
         addPathAnnotation(methodCreator,
                 appendToPath(resourceProperties.getPath(RESOURCE_UPDATE_METHOD_NAME), "{id}"));
-        addTransactionalAnnotation(methodCreator);
         addPutAnnotation(methodCreator);
         addPathParamAnnotation(methodCreator.getParameterAnnotations(0), "id");
         addConsumesAnnotation(methodCreator, APPLICATION_JSON);
         addProducesAnnotation(methodCreator, APPLICATION_HAL_JSON);
 
-        // Invoke resource methods
         ResultHandle resource = methodCreator.readInstanceField(resourceField, methodCreator.getThis());
         ResultHandle id = methodCreator.getMethodParam(0);
         ResultHandle entityToSave = methodCreator.getMethodParam(1);
 
-        // Wrap and return response
-        BranchResult entityExists = doesEntityExist(methodCreator, resourceMetadata.getResourceClass(), resource, id);
+        // Invoke resource methods
+        TryBlock tryBlock = implementTryBlock(methodCreator, "Failed to update an entity");
+        BranchResult entityExists = doesEntityExist(tryBlock, resourceMetadata.getResourceClass(), resource, id);
         updateAndReturn(entityExists.trueBranch(), resourceMetadata.getResourceClass(), resource, id, entityToSave);
         createAndReturn(entityExists.falseBranch(), resourceMetadata.getResourceClass(), resource, id, entityToSave);
+
+        tryBlock.close();
         methodCreator.close();
     }
 

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResponseImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResponseImplementor.java
@@ -55,6 +55,10 @@ public final class ResponseImplementor {
         return status(creator, Response.Status.NO_CONTENT.getStatusCode());
     }
 
+    public static ResultHandle notFound(BytecodeCreator creator) {
+        return status(creator, Response.Status.NOT_FOUND.getStatusCode());
+    }
+
     public static ResultHandle notFoundException(BytecodeCreator creator) {
         return creator.newInstance(MethodDescriptor.ofConstructor(WebApplicationException.class, int.class),
                 creator.load(Response.Status.NOT_FOUND.getStatusCode()));

--- a/extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/RestDataPanacheException.java
+++ b/extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/RestDataPanacheException.java
@@ -1,0 +1,8 @@
+package io.quarkus.rest.data.panache;
+
+public class RestDataPanacheException extends RuntimeException {
+
+    public RestDataPanacheException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}

